### PR TITLE
[FIX] Passing checks

### PR DESCRIPTION
**Pull request by ravinkohli**
This PR fixes the following 

-  tests which were failing due to the updated implementation in the branch.
-  In addition as certain components (preprocessors and network embedding) require a fix before tests can be run on them, I have currently skipped those tests.
- mypy and flake checks

The goal of this PR is to pass all checks for the reg_cocktails branch.

### CODE REVIEW

File: autoPyTorch/data/tabular_feature_validator.py [hunk 1]
```diff
@@ -246,13 +247,12 @@ def transform(
...
-            for column in categorical_columns:
-                if X[column].isna().all():
-                    X[column] = X[column].astype('object')
-
         if self.column_transformer is not None:
+            if len(self.categorical_columns) > 0:
+                categorical_columns = self.column_transformer.transformers_[0][-1]
+                for column in categorical_columns:
+                    if X[column].isna().all():
+                        X[column] = X[column].astype('object')
```
[the remark this answers is not in the document]
Comment (ravinkohli) [reply autoPyTorch/data/tabular_feature_validator.py:255]: no as `all_nan_columns` are the nan columns of the `X_train`, `transform` can be called also on test or val data

File: autoPyTorch/data/tabular_feature_validator.py [hunk 2]
```diff
@@ -345,11 +345,16 @@ def _check_data(
...
                 and len(set(X.columns[dtypes_diff]).difference(self.all_nan_columns)) != 0
             ):
-                # The dtypes can be different if and only if the column belongs
-                # to all_nan_columns as these columns would be imputed.
-                raise ValueError("The dtype of the features must not be changed after fit(), but"
-                                 " the dtypes of some columns are different between training ({}) and"
-                                 " test ({}) datasets.".format(self.dtypes, dtypes))
+                diff_indices = np.where(np.array(dtypes_diff))[0]
+
+                if not all([X[X.columns[i]].isna().all() for i in diff_indices]):
```
[the remark this answers is not in the document]
Comment (ravinkohli) [reply autoPyTorch/data/tabular_feature_validator.py:350]: no as mentioned in the previous comment `all_nan_columns` are the nan columns of the `X_train`, `transform` can be called also on test or val data

File: autoPyTorch/pipeline/components/training/trainer/AdversarialTrainer.py [hunk 3]
```diff
@@ -109,11 +109,7 @@ def train_step(self, data: np.ndarray, targets: np.ndarray) -> Tuple[float, torc
         loss = loss_func(self.criterion, original_outputs, adversarial_outputs)
         loss.backward()
         self.optimizer.step()
-        if self.scheduler:
-            if 'ReduceLROnPlateau' in self.scheduler.__class__.__name__:
-                self.scheduler.step(loss)
-            else:
-                self.scheduler.step()
```
[the remark this answers is not in the document]
Comment (ravinkohli) [reply autoPyTorch/pipeline/components/training/trainer/AdversarialTrainer.py:116]: because for now we only update the step in `train_epoch` of the `BaseTrainerComponent`. These lines were a bug as for `AdversarialTrainer` the scheduler is updated every batch as well as epoch

File: autoPyTorch/data/tabular_feature_validator.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nabenabe0928) [reply autoPyTorch/data/tabular_feature_validator.py:349]: Wait, I would like to check a few points:

First, the following understanding is correct?
Suppose some columns in train and test datasets have different dtypes,
1. if both train and test datasets have all NaN, it does not matter ==> both float
2. If one of train or test datasets have all NaN, it does not matter ==> one float and the other any
3. If both train and test datasets do not have all NaN, it might matter ==> for example, float in train and str in test because a user wrongly mapped training dataset, which does not include leading zero, and test dataset has some numbers that are leading zero.

Then, we should raise an error only when the point 3 happens, right?
I think we wrongly allow some of point 3 conditions and filter some of point 2 condition.

The precise condition here should be something like this, isn't it?:
```
    def is_datasets_consistent(self, diff_cols: List[str]) -> bool:
        """
        Check the consistency of dtypes between training and test datasets.

        Args:
            diff_cols (List[bool]): The column labels that have different dtypes.

        Returns:
            _ (bool): Whether the training and test datasets are consistent.
        """

        if self.all_nan_columns is None:
            return True

        # dtype is different ==> the column in at least either of train or test datasets must be all NaN
        # inconsistent ==> dtype is different and the col in both train and test is not all NaN
        inconsistent_cols = list(set(diff_cols) - self.all_nan_columns)
        
        return len(inconsistent_cols) == 0 or all(X[inconsistent_cols].isna().all())
```

**Review by nabenabe0928 — CHANGES_REQUESTED** [review]

File: autoPyTorch/data/tabular_feature_validator.py [hunk 4]
```diff
@@ -345,11 +344,16 @@ def _check_data(
...
+                    # The dtypes can be different if the column belongs
+                    # to all_nan_columns if the dataset is in training
+                    # set or if the column is all nan if the dataset
+                    # is for testing as these columns would be imputed.
+                    raise ValueError("The dtype of the features must not be changed after fit(), but"
+                                     " the dtypes of some columns are different between training ({}) and"
+                                     " test ({}) datasets.".format(self.dtypes, dtypes))
```
Comment (nabenabe0928) [inline autoPyTorch/data/tabular_feature_validator.py:357]:
```suggestion

    def _is_datasets_consistent(self, diff_cols: List[str]) -> bool:
        """
        Check the consistency of dtypes between training and test datasets.

        Args:
            diff_cols (List[bool]): The column labels that have different dtypes.

        Returns:
            _ (bool): Whether the training and test datasets are consistent.
        """

        if self.all_nan_columns is None:
            return True

        # dtype is different ==> the column in at least either of train or test datasets must be all NaN
        # inconsistent ==> dtype is different and the col in both train and test is not all NaN
        inconsistent_cols = list(set(diff_cols) - self.all_nan_columns)
        
        return len(inconsistent_cols) == 0 or all(X[inconsistent_cols].isna().all())
        
```

File: autoPyTorch/data/tabular_feature_validator.py [hunk 5]
```diff
@@ -345,11 +344,16 @@ def _check_data(
                 and self.all_nan_columns is not None  # Ignore all_nan_columns is None
```
Comment (nabenabe0928) [inline autoPyTorch/data/tabular_feature_validator.py:344]: So the new version will be this:
```
            diff_cols = X.columns[[s_dtype != dtype for s_dtype, dtype in zip(self.dtypes, dtypes)]]
            if len(self.dtypes) == 0:
                self.dtypes = dtypes
            elif not self._is_datasets_consistent(diff_cols):
                # The dtypes can be different if the column belongs
                # to all_nan_columns if the dataset is in training
                # set or if the column is all nan if the dataset
                # is for testing as these columns would be imputed.
                raise ValueError("The dtype of the features must not be changed after fit(), but"
                                 " the dtypes of some columns are different between training ({}) and"
                                 " test ({}) datasets.".format(self.dtypes, dtypes))
```

File: test/test_data/test_feature_validator.py [hunk 6]
```diff
@@ -237,9 +236,10 @@ def test_featurevalidator_categorical_nan(input_data_featuretest):
     validator.fit(input_data_featuretest)
     transformed_X = validator.transform(input_data_featuretest)
     assert any(pd.isna(input_data_featuretest))
-    assert any((-1 in categories) or ('-1' in categories) or ('Missing!' in categories) for categories in
-               validator.encoder.named_transformers_['encoder'].categories_)
-    assert np.shape(input_data_featuretest) == np.shape(transformed_X)
```
[the remark this answers is not in the document]
Comment (ravinkohli) [reply test/test_data/test_feature_validator.py:242]: because we are using a one hot encoder

File: autoPyTorch/data/tabular_feature_validator.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ravinkohli) [reply autoPyTorch/data/tabular_feature_validator.py:349]: yes you are right. I'll change this.

File: autoPyTorch/data/tabular_feature_validator.py (hunk 4 above)
Comment (ravinkohli) [reply autoPyTorch/data/tabular_feature_validator.py:357]: okay it also makes it easier to understand. I will commit this.

File: autoPyTorch/data/tabular_feature_validator.py [hunk 7]
```diff
@@ -508,6 +506,23 @@ def infer_objects(self, X: pd.DataFrame) -> pd.DataFrame:
 
         return X
 
+    def _is_datasets_consistent(self, diff_cols: List[str], X: pd.DataFrame) -> bool:
+        """
+        Check the consistency of dtypes between training and test datasets.
+        Args:
+            diff_cols (List[bool]): The column labels that have different dtypes.
+        Returns:
+            _ (bool): Whether the training and test datasets are consistent.
```
Comment (nabenabe0928) [inline autoPyTorch/data/tabular_feature_validator.py:515]: Sorry I forgot adding `X`

```suggestion
        Check the consistency of dtypes between training and test datasets.

        Args:
            diff_cols (List[bool]): The column labels that have different types.
            X (pd.DataFrame): A validation or test dataset to be compared with the training dataset

        Returns:
            _ (bool): Whether the training and test datasets are consistent.
```

File: autoPyTorch/data/tabular_feature_validator.py [hunk 8]
```diff
@@ -508,6 +506,23 @@ def infer_objects(self, X: pd.DataFrame) -> pd.DataFrame:
...
+        """
+        if self.all_nan_columns is None:
+            return True
+
+        # dtype is different ==> the column in at least either of train or test datasets must be all NaN
+        # inconsistent <==> dtype is different and the col in both train and test is not all NaN
+        inconsistent_cols = list(set(diff_cols) - self.all_nan_columns)
+
+        return len(inconsistent_cols) == 0 or all(X[inconsistent_cols].isna().all())
```
Comment (nabenabe0928) [inline autoPyTorch/data/tabular_feature_validator.py:524]: Which is better? You can take according to your preference.
```suggestion
        return len(inconsistent_cols) == 0 or X[inconsistent_cols].isna().all().all()
```

**Review by nabenabe0928 — COMMENTED** [review]
Thanks for addressing my comments.
I noticed that my comments was not completed properly, so I added comments.

File: autoPyTorch/data/tabular_feature_validator.py [hunk 9]
```diff
@@ -337,16 +337,14 @@ def _check_data(
...
-                and self.all_nan_columns is not None  # Ignore all_nan_columns is None
-                and len(set(X.columns[dtypes_diff]).difference(self.all_nan_columns)) != 0
-            ):
-                # The dtypes can be different if and only if the column belongs
-                # to all_nan_columns as these columns would be imputed.
+            elif not self._is_datasets_consistent(diff_cols, X):
+                # The dtypes can be different if the column belongs
+                # to all_nan_columns if the dataset is in training
+                # set or if the column is all nan if the dataset
+                # is for testing as these columns would be imputed.
```
Comment (nabenabe0928) [inline autoPyTorch/data/tabular_feature_validator.py:347]: Probably, we can delete those comments or move them to the doc-string of `is_datasets_consistent`?

Comment (ravinkohli) [reply autoPyTorch/data/tabular_feature_validator.py:347]: yes I was thinking the same thing. But I wanted to wait for the tests to pass. I'll change this if they pass now.

File: autoPyTorch/data/tabular_feature_validator.py [hunk 10]
```diff
@@ -246,13 +247,12 @@ def transform(
         # having a value for a categorical column.
         # We need to convert the column in test data to
         # object otherwise the test column is interpreted as float
-        if len(self.categorical_columns) > 0:
-            categorical_columns = self.column_transformer.transformers_[0][-1]
-            for column in categorical_columns:
-                if X[column].isna().all():
-                    X[column] = X[column].astype('object')
-
         if self.column_transformer is not None:
```
Comment (ArlindKadra) [inline autoPyTorch/data/tabular_feature_validator.py:250]: Is this just to pass the tests, theoretically this is always instantiated during `_fit` right

Comment (ravinkohli) [reply autoPyTorch/data/tabular_feature_validator.py:250]: So, the lines moved into this condition that is to make columns 'object' dtype, require that the data `X` is a dataframe. If there is a column transformer,  we are sure that the data is a dataframe.

File: autoPyTorch/data/tabular_feature_validator.py [hunk 11]
```diff
@@ -508,6 +502,30 @@ def infer_objects(self, X: pd.DataFrame) -> pd.DataFrame:
...
+
+        Args:
+            diff_cols (List[bool]):
+                The column labels that have different dtypes.
+            X (pd.DataFrame):
+                A validation or test dataset to be compared with the training dataset
+        Returns:
+            _ (bool): Whether the training and test datasets are consistent.
+        """
+        if self.all_nan_columns is None:
```
Comment (ArlindKadra) [inline autoPyTorch/data/tabular_feature_validator.py:520]: when `self.all_nan_columns`, should you not check only for the second part at line 527?
it does not seem like it should return true, it seems like `inconsistent_cols` should be `diff_cols`.

**Review by ArlindKadra — COMMENTED** [review]
Looks good, although I have a few uncertainties :)

File: autoPyTorch/data/tabular_feature_validator.py (hunk 10 above)
Comment (ravinkohli) [reply autoPyTorch/data/tabular_feature_validator.py:250]: we technically also support sparse data. So the test for sparse data were failing

File: autoPyTorch/pipeline/components/setup/network_backbone/InceptionTimeBackbone.py [hunk 12]
```diff
@@ -78,7 +78,7 @@ def __init__(self, n_res_inputs: int, n_outputs: int):
     def forward(self, x: torch.Tensor, res: torch.Tensor) -> torch.Tensor:
         shortcut = self.shortcut(res)
         shortcut = self.bn(shortcut)
-        x += shortcut
+        x = x + shortcut
```
[the remark this answers is not in the document]
Comment (ravinkohli) [reply autoPyTorch/pipeline/components/setup/network_backbone/InceptionTimeBackbone.py:81]: no it was causing an error

File: autoPyTorch/data/tabular_feature_validator.py (hunk 11 above)
Comment (ravinkohli) [reply autoPyTorch/data/tabular_feature_validator.py:520]: oh thats true, let me fix that

File: autoPyTorch/api/base_task.py [hunk 13]
```diff
@@ -1429,6 +1429,7 @@ def fit_ensemble(
                              "Please call the `search()` method of {} prior to "
                              "fit_ensemble().".format(self.__class__.__name__))
 
+        precision = precision if precision is not None else self.precision
```
[the remark this answers is not in the document]
Comment (ravinkohli) [reply autoPyTorch/api/base_task.py:1432]: yeah true, I'll change this as well

File: autoPyTorch/pipeline/components/setup/network_backbone/InceptionTimeBackbone.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (ArlindKadra) [reply autoPyTorch/pipeline/components/setup/network_backbone/InceptionTimeBackbone.py:81]: surprising, why is there a different behavior.

File: autoPyTorch/data/tabular_feature_validator.py (hunk 10 above)
Comment (ArlindKadra) [reply autoPyTorch/data/tabular_feature_validator.py:250]: Ah, then it is understandable.

File: autoPyTorch/data/tabular_feature_validator.py (hunk 11 above)
Comment (ArlindKadra) [reply autoPyTorch/data/tabular_feature_validator.py:520]: I think this still does not fix it because it can be that `diff_cols` is not empty, but they are all `None`. So, the second `or `part of the return.

File: autoPyTorch/pipeline/components/setup/network_backbone/InceptionTimeBackbone.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (ravinkohli) [reply autoPyTorch/pipeline/components/setup/network_backbone/InceptionTimeBackbone.py:81]: I think its due to a new torch version

File: autoPyTorch/data/tabular_feature_validator.py (hunk 11 above)
Comment (ArlindKadra) [reply autoPyTorch/data/tabular_feature_validator.py:520]:
```
 if self.all_nan_columns is None:
     if len(diff_cols) == 0:
         return True
     else:
         return False
```
If `self.nan_columns` is `None`, you check if there is a difference in the types by `len(diff_cols)`, if there is, you return `False`. Should you not also consider the case that there might be a difference in the types, however all of those columns might be NA fully for test?

`return len(inconsistent_cols) == 0 or all(X[inconsistent_cols].isna().all())` this is done by the second part of this or with 2 clauses. You are not reaching here and it might be that it is this case.

Maybe I am missunderstanding something?

File: test/test_data/test_feature_validator.py [hunk 14]
```diff
@@ -237,9 +236,10 @@ def test_featurevalidator_categorical_nan(input_data_featuretest):
...
+    categories_ = validator.column_transformer.\
+        named_transformers_['categorical_pipeline'].named_steps['onehotencoder'].categories_
+    assert any(('0' in categories) or (0 in categories) or ('missing_value' in categories) for categories in
```
Comment (ArlindKadra) [inline test/test_data/test_feature_validator.py:241]: This are the unique category values for the one hot encoder right, what would be the case where they are 0, is it not 'missing_value' for categorical columns?

File: autoPyTorch/data/tabular_feature_validator.py (hunk 11 above)
Comment (ravinkohli) [reply autoPyTorch/data/tabular_feature_validator.py:520]: okay I got it, yeah it makes sense

File: test/test_data/test_feature_validator.py (hunk 14 above)
Comment (ArlindKadra) [reply test/test_data/test_feature_validator.py:241]: according to this https://github.com/automl/Auto-PyTorch/blob/10aea660389c899e9ab16410e95371c7c97f2436/autoPyTorch/data/tabular_feature_validator.py#L380-L381

It will be a categorical column only when it has a `string` or `bool` dtype.
And if between test and train there are differences, it will either throw an error or convert a null column to object for test if not empty in train.

Comment (ArlindKadra) [reply test/test_data/test_feature_validator.py:241]: Does it fail if you run it without the `0` checks?

Comment (ravinkohli) [reply test/test_data/test_feature_validator.py:241]: > It will be a categorical column only when it has a `string` or `bool` dtype

not string, but category. And category can be string or int.

Comment (ravinkohli) [reply test/test_data/test_feature_validator.py:241]: It works fine without `'0'` check but removing `0` gives an error
